### PR TITLE
HDDS-5919. In kubernetes OM HA has circular dependency on the service availability

### PR DESCRIPTION
**Pull request by sokui**
## What changes were proposed in this pull request?

1. we can control to disable JVM network address cache to be enabled or not2. 
2. OM can use hostname (ozone-om-0) instead of fqdn (ozone-om-o.ozone-om-service) to resolve itself's hostname during the launching time.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-5919

## How was this patch tested?
Test it in k8s production with kerberos enabled. works well. In the launching time OM can resolve itself with the hostname.

### CODE REVIEW

**Comment by kerneltime** [comment]
cc @sodonnel

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/FlexibleFQDNResolution.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.security.Security;
+
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_JVM_NETWORK_ADDRESS_CACHE_ENABLED;
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_JVM_NETWORK_ADDRESS_CACHE_ENABLED_DEFAULT;
+
+/**
+ * FQDN related utils.
```
Comment (kerneltime) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/FlexibleFQDNResolution.java:31]: Please add more context as to why this is needed.

File: hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/ha/TestFlexibleFQDNResolution.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.hadoop.net.NetUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.net.InetSocketAddress;
+
+/**
+ * Tests for {@link FlexibleFQDNResolution} class.
+ */
+public class TestFlexibleFQDNResolution {
```
Comment (kerneltime) [inline hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/ha/TestFlexibleFQDNResolution.java:28]: Would this work only in k8s environment?

Comment (sokui) [reply hadoop-hdds/common/src/test/java/org/apache/hadoop/ozone/ha/TestFlexibleFQDNResolution.java:28]: By turning on ozone.flexible.fqdn.resolution.enabled, it should work for both traditional servers and k8s based servers. Of course, the traditional servers do not require it. You can turn it off when deploying to the traditional servers

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/FlexibleFQDNResolution.java (hunk 1 above)
Comment (sokui) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/FlexibleFQDNResolution.java:31]: will do . Thanks

**Review by adoroszlai — COMMENTED** [review]
Thanks @sokui for working on this.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java [hunk 3]
```diff
@@ -448,6 +448,16 @@ public final class OzoneConfigKeys {
...
+  public static final String OZONE_FLEXIBLE_FQDN_RESOLUTION_ENABLED =
+          "ozone.flexible.fqdn.resolution.enabled";
+  public static final boolean OZONE_FLEXIBLE_FQDN_RESOLUTION_ENABLED_DEFAULT =
+          false;
+
+  public static final String OZONE_JVM_NETWORK_ADDRESS_CACHE_ENABLED =
+          "ozone.jvm.network.address.cache.enabled";
+  public static final boolean OZONE_JVM_NETWORK_ADDRESS_CACHE_ENABLED_DEFAULT =
+          true;
+
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java:460]: updated

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/ConfUtils.java [hunk 4]
```diff
@@ -69,6 +69,9 @@ public static String concatSuffixes(String... suffixes) {
    * Return true if it matches, false otherwsie.
    */
   public static boolean isAddressLocal(InetSocketAddress addr) {
+    if (addr.getAddress() == null) {
+      return false;
+    }
     return NetUtils.isLocalAddress(addr.getAddress());
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/ConfUtils.java:75]: updated

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHANodeDetails.java [hunk 5]
```diff
@@ -227,13 +230,22 @@ public static SCMHANodeDetails loadSCMHAConfig(OzoneConfiguration conf)
           throw e;
         }
 
-        if (addr.isUnresolved()) {
+        boolean flexibleFqdnResolutionEnabled = conf.getBoolean(
+                OZONE_FLEXIBLE_FQDN_RESOLUTION_ENABLED,
+                OZONE_FLEXIBLE_FQDN_RESOLUTION_ENABLED_DEFAULT);
+        if (!flexibleFqdnResolutionEnabled && addr.isUnresolved()
+                || flexibleFqdnResolutionEnabled
+                && !isAddressHostNameLocal(addr)) {
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHANodeDetails.java:238]: updated

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHANodeDetails.java [hunk 6]
```diff
@@ -227,13 +230,22 @@ public static SCMHANodeDetails loadSCMHAConfig(OzoneConfiguration conf)
...
                   + "with unresolved host to create Ratis ring.", nodeId,
               rpcAddrStr);
         }
 
-        if (!addr.isUnresolved() && !isPeer && ConfUtils.isAddressLocal(addr)) {
+        if (!isPeer && (!flexibleFqdnResolutionEnabled
+                && !addr.isUnresolved()
+                && ConfUtils.isAddressLocal(addr)
+                || flexibleFqdnResolutionEnabled
+                && isAddressHostNameLocal(addr))) {
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHANodeDetails.java:248]: updated

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/FlexibleFQDNResolution.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_JVM_NETWORK_ADDRESS_CACHE_ENABLED;
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_JVM_NETWORK_ADDRESS_CACHE_ENABLED_DEFAULT;
+
+/**
+ * FQDN related utils. Provides methods related to local host name and
+ * jvm network address cache. This utils is mainly used in kubernetes where
+ * the FQDN [pod_name].[service_name] is not resolvable at the service
+ * starting time.
+ */
+public final class FlexibleFQDNResolution {
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/FlexibleFQDNResolution.java:36]: updated

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java [hunk 8]
```diff
@@ -151,6 +152,8 @@ private static HddsDatanodeService createHddsDatanodeService(
 
   public static void main(String[] args) {
     try {
+      FlexibleFQDNResolution.disableJvmNetworkAddressCacheIfRequired(
+              new OzoneConfiguration());
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java:156]: added this to both S3G and Recon

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerStarter.java [hunk 9]
```diff
@@ -51,6 +52,8 @@ public class OzoneManagerStarter extends GenericCli {
       LoggerFactory.getLogger(OzoneManagerStarter.class);
 
   public static void main(String[] args) throws Exception {
+    FlexibleFQDNResolution.disableJvmNetworkAddressCacheIfRequired(
+            new OzoneConfiguration());
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerStarter.java:56]: yeah. I think I tried that before. It is too late

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManagerStarter.java [hunk 10]
```diff
@@ -56,6 +57,8 @@ public class StorageContainerManagerStarter extends GenericCli {
       LoggerFactory.getLogger(StorageContainerManagerStarter.class);
 
   public static void main(String[] args) {
+    FlexibleFQDNResolution.disableJvmNetworkAddressCacheIfRequired(
+            new OzoneConfiguration());
```
[the remark this answers is not in the document]
Comment (sokui) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManagerStarter.java:61]: same as above

**Comment by adoroszlai** [comment]
```
hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ha/ConfUtils.java
 23: Unused import - org.apache.hadoop.net.NetUtils.
 26: Unused import - java.net.InetAddress.
 27: Unused import - java.net.InetSocketAddress.
```

**Comment by adoroszlai** [comment]
Thanks @sokui for updating the patch.  LGTM.

**Comment by adoroszlai** [comment]
@GeorgeJahad @xBis7 can you please take a look at this Kubernetes-related fix?

**Comment by GeorgeJahad** [comment]
I'll have some time later today @adoroszlai

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java [hunk 11]
```diff
@@ -262,7 +262,7 @@ private static PKCS10CertificationRequest generateCSR(
         client.getPrivateKey());
 
     // Get host name.
-    String hostname = scmAddress.getAddress().getHostName();
+    String hostname = scmAddress.getHostName();
```
Comment (GeorgeJahad) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java:265]: Is this change necessary?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/util/OzoneNetUtils.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  /**
+   * For the input FQDN address, return a new address with its host name
+   * (without the domain name) and port.
+   *
+   * @param addr a FQDN address
+   * @return The address of host name
+   */
+  public static InetSocketAddress getAddressWithHostName(
```
Comment (GeorgeJahad) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/util/OzoneNetUtils.java:80]: Nit:  for me, the OM changes would be easier to understand if this method were called "getAddressWithHostNameLocal()", (which would parallel the "isAddressHostNameLocal()" method above.)

Comment (GeorgeJahad) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/util/OzoneNetUtils.java:80]: lgtm

**Comment by GeorgeJahad** [comment]
lgtm

**Comment by adoroszlai** [comment]
@kerneltime would you like to take another look?
@JacksonYao287 would you like to take a look?

If not, I would like to merge this.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java (hunk 11 above)
Comment (sokui) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java:265]: I cannot exactly remember. I think I got some cert issue when deploying to k8s. But when I just tested this, it seems both of these correctly return the hostname. I cannot remember now why I have this change. Are you OK to keep my change, or you think I should revert it? Please let me know. Thanks

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/util/OzoneNetUtils.java (hunk 12 above)
Comment (sokui) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/util/OzoneNetUtils.java]: updated

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java (hunk 11 above)
Comment (GeorgeJahad) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java:265]: I'm ok with keeping it, I think it does the same thing.  I was just wondering.

**Comment by JacksonYao287** [comment]
LGTM +1

**Comment by adoroszlai** [comment]
Thanks @sokui for the patch, @GeorgeJahad, @JacksonYao287, @kerneltime for the review.
